Add `touch` option to `#update_columns` and `#update_column` methods

moofkit opened the pull request:
<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), the motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->
I've created this Pull Request because not all ActiveRecord persistence methods affect timestamps or have a touch option. This can be problematic, especially when there is some ETL processing that relies on the `updated_at` timestamps instead of copying the whole table. For instance, if someone uses `#update_columns` for performance reasons, it may result in lost updates. This is particularly an issue for `#update_column`. If a user wants to keep timestamps current, they have to explicitly call the `#touch` method after using `#update_column`.

Examples:
```ruby
# we don't want to perform any callbacks or validations here so use #update_columns
current_user.update_columns(
  last_ip: request.remote_ip,
  updated_at: Time.current # but we still want to keep track of the last changes, so have to provide a timestamp explicitly
)
```

```ruby
current_user.update_column(:last_ip, request.remote_ip)
current_user.touch
```

Rails discussion: https://discuss.rubyonrails.org/t/proposal-add-touch-option-for-update-columns-update-column/85388

### Detail

This PR adds a `touch` option to the `#update_columns` and `#update_column` methods.
The option allows to update `updated_at`/`updated_on` attributes with 

```ruby
    person.update_columns(name: 'John Doe', touch: true)
```

### Additional information

<!-- Provide additional information such as benchmarks, references to other repositories, or alternative solutions. -->

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

fatkodima commented on activemodel/CHANGELOG.md:
```diff
@@ -19,4 +19,9 @@
 
     *Jonathan Hefner*
 
+*   Add `touch` option to `#update_columns` and `#update_column` methods. When
```
This should be in `activerecord/CHANGELOG.md`.

I would change it to something like
```
*  Add `touch` option to `#update_columns` and `#update_column` methods.

   Example:
   ```ruby
   # Will update :updated_at/:updated_on alongside :nice column.
   user.update_column(:nice, true, touch: true)
   ```
```

fatkodima commented on activerecord/lib/active_record/persistence.rb:
```diff
@@ -916,26 +916,35 @@ def update!(attributes)
     # * \Validations are skipped.
     # * \Callbacks are skipped.
-    # * +updated_at+/+updated_on+ are not updated.
+    # * +updated_at+/+updated_on+ are updated if the +touch+ option is set to +true+. Otherwise, they're not updated.
```
```suggestion
    # * +updated_at+/+updated_on+ are updated if the +touch+ option is set to +true+.
```

fatkodima commented on activerecord/lib/active_record/persistence.rb:
```diff
@@ -916,26 +916,35 @@ def update!(attributes)
+    #
+    # ==== Parameters
+    # * <tt>:touch</tt> option - Touch the timestamp columns when updating.
+    # * If attributes names are passed, they are updated along with update_at/on attributes.
```
```suggestion
    # * If attribute names are passed, they are updated along with +updated_at+/+updated_on+ attributes.
```

fatkodima commented on activerecord/lib/active_record/persistence.rb:
```diff
@@ -947,6 +956,15 @@ def update_columns(attributes)
+        names = Array.wrap(names)
+        options = names.extract_options!
+        touch_updates = self.class.touch_attributes_with_time(*names, **options)
+        attributes.reverse_merge!(touch_updates) if touch_updates.any?
```
```suggestion
        attributes.reverse_merge!(touch_updates) unless touch_updates.empty?
```

fatkodima commented on activerecord/test/cases/persistence_test.rb:
```diff
@@ -1077,6 +1077,15 @@ def test_update_column
 
+  def test_update_column_touch_option
+    topic = Topic.find(1)
+    updated_at = topic.updated_at
+    travel(1.second) do
+      topic.update_column(:title, "super_title", touch: true)
+    end
+    assert_not_equal updated_at, topic.updated_at
```
I think, in this and the following tests we can utilise the `assert_changes` test helper.

moofkit committed:
Add `touch` option to `#update_columns` and `#update_column` methods

    This PR adds a `touch` option to the `#update_columns` and `#update_column` methods. This option allows you to update the `updated_at` timestamp of the record without running callbacks or validations.

        person.update_columns(name: 'John Doe', touch: true)

moofkit commented:
@fatkodima, thank you for the review and suggestions! I have implemented all of them and rebased the branch. Please let me know if you have any further suggestions.

p8 commented:
This change would break existing applications that have models with a `touch` column.

willianveiga commented:
> This change would break existing applications that have models with a `touch` column.

I don't think it's possible:
```bash
willian@veiga:/tmp/foo$ rails generate scaffold bar touch:string
Could not generate field 'touch', as it is already defined by Active Record.
```

willianveiga commented:
> This change would break existing applications that have models with a `touch` column.

I don't think it's possible:
```bash
$ rails generate scaffold bar touch:string
Could not generate field 'touch', as it is already defined by Active Record.
```

moofkit commented:
> This change would break existing applications that have models with a `touch` column.

@p8 thanks for pointing this out! It's possible for sure.
But there are several ambiguous column names that it's recommended to avoid. Like `save`, and `touch` because those names are reserved by the active record API.
For example, if a user defines the `touch` column, the `#touch` [method](https://api.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-touch)  will be overridden by attribute accessor, which might be a more serious problem.
By the way, Rubocop has a [cop](https://docs.rubocop.org/rubocop-rails/cops_rails.html#railsdangerouscolumnnames) for this.

moofkit commented:
> > This change would break existing applications that have models with a `touch` column.
> 
> I don't think it's possible:
> 
> ```shell
> $ rails generate scaffold bar touch:string
> Could not generate field 'touch', as it is already defined by Active Record.
> ```

I have checked the rails master and it raises an error even with forced migration. Thanks, @willianveiga for the notice!
This is a reproducible example
```ruby
# touch_field_error.rb
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "rails", github: "rails/rails", branch: "main"
  gem "sqlite3", "~> 1.4"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :comments, force: true do |t|
    t.boolean :touch
  end
end

class Comment < ActiveRecord::Base
end

class Test < Minitest::Test
  def test_association_stuff
    comment = Comment.create!
  end
end
```

```bash
$ ruby touch_field_error.rb
# Running:

E

Finished in 0.002023s, 494.3153 runs/s, 0.0000 assertions/s.

  1) Error:
Test#test_association_stuff:
ActiveRecord::DangerousAttributeError: touch is defined by Active Record. Check to make sure that you don't have an attribute or method with the same name.
```

So using the `touch` field as an model attribute seems problematic and dangerous, even without changes introduced by this PR.

@p8 or I'm missing something?

clemensp commented:
Hello, what are the next steps for this feature? I'm currently in a situation where I have a handful of update_column/update_columns calls that leave the updated_at out of sync, causing problems when trying to fetch recently updated records.

I could look at triggering timestamp updates manually, but that wouldn't guarantee that I capture all current and future call sites of the methods that eventually call update_column(s).

This looks like the ideal approach.

rafaelfranca commented on activerecord/lib/active_record/persistence.rb:
```diff
@@ -613,6 +627,15 @@ def update_columns(attributes)
+      if touch
+        names = touch if touch != true
+        names = Array.wrap(names)
+        options = names.extract_options!
```
Why `extract_options!` here? I don't see any option being passed in the tests.

rafaelfranca replied:
But there is no test coverage for that, so if we remove this line no test will fail. We should add.

fatkodima replied:
I will add for this and other methods. Let me know if you want to do by yourself.

rafaelfranca replied:
Please do it. I'm still catching up
